src/Functions/parseTimeDelta.cpp

```diff
@@ -0,0 +1,274 @@
+     *
+     * The length of years and months (and even days in presence of time adjustments) are rough:
+     * year is just 365 days, month is 30.5 days, day is 86400 seconds, similarly to what formatReadableTimeDelta is doing.
+     */
+    class FunctionParseTimeDelta : public IFunction
+    {
+    public:
+        static constexpr auto name = "parseTimeDelta";
+        static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionParseTimeDelta>(); }
+
+        String getName() const override { return name; }
+
+        bool isVariadic() const override { return true; }
+
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+        size_t getNumberOfArguments() const override { return 0; }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+        {
+            if (arguments.empty())
+                throw Exception(
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    toString(arguments.size()));
+
+            if (arguments.size() > 1)
+                throw Exception(
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    toString(arguments.size()));
+
+            const IDataType & type = *arguments[0];
+
+            if (!isString(type))
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Cannot format {} as time string.", type.getName());
+
+            return std::make_shared<DataTypeFloat64>();
+        }
+
+        bool useDefaultImplementationForConstants() const override { return true; }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            auto col_to = ColumnFloat64::create();
+            auto & res_data = col_to->getData();
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                std::string_view str{arguments[0].column->getDataAt(i)};
+                Int64 token_tail = 0;
+                Int64 token_front = 0;
+                Int64 last_pos = str.length() - 1;
+                Float64 result = 0;
+
+                /// ignore '.' and ' ' at the end of string
+                while (last_pos >= 0 && (str[last_pos] == ' ' || str[last_pos] == '.'))
+                    --last_pos;
+
+                /// no valid characters
+                if (last_pos < 0)
+                {
+                    res_data.emplace_back(0);
+                    continue;
+                }
+
+                while (token_tail <= last_pos)
+                {
+                    scanSeparator(str, token_tail, last_pos);
```

Comment:
Maybe don't allow separators before first unit?
Like `parseTimeDelta(':1h')`
What do you think?

Reply:
I thought about it here. At that time, I felt that both ways were fine, and I chose the simpler one.
Now I think it's better not to allow separators before first unit and I will fix it as soon.